### PR TITLE
GEODE-7601: Memory leaks in C++ native client found when running old integration-tests

### PR DESCRIPTION
**Pull request by albertogpz**
Several memory leaks and other issues in memory
management have been fixed in the native client.
The issues were found when running the old integration
tests under valgrind.

Many issues in the testing framework have also been fixed
to make it easier to identify and fix the problems
in the native client.

### CODE REVIEW

File: cppcache/include/geode/DataOutput.hpp [hunk 1]
```diff
@@ -497,8 +497,12 @@ class APACHE_GEODE_EXPORT DataOutput {
   static void acquireLock();
   static void releaseLock();
 
+  struct FreeDeleter {
+    void operator()(uint8_t* p) { free(p); }
+  };
+
   // memory m_buffer to encode to.
-  std::unique_ptr<uint8_t[]> m_bytes;
+  std::unique_ptr<uint8_t, FreeDeleter> m_bytes;
```
Comment (pivotal-jbarrett) [inline cppcache/include/geode/DataOutput.hpp]: What would it look like to convert this to a `std::vector`? Then we get all this for free.

File: cppcache/integration-test/ThinClientFailover3.hpp [hunk 2]
```diff
@@ -242,21 +242,22 @@ void doNetsearch(const char* name, const char* key, const char* value) {
...
 }
 
-std::vector<char*> storeEndPoints(const char* points) {
-  std::vector<char*> endpointNames;
-  if (points != nullptr) {
-    char* ep = strdup(points);
+std::vector<std::string> storeEndPoints(const std::string points) {
+  std::vector<std::string> endpointNames;
+  if (!points.empty()) {
+    char* ep = strdup(points.c_str());
```
Comment (pivotal-jbarrett) [inline cppcache/integration-test/ThinClientFailover3.hpp]: Maybe we can keep this all as `std::string`?

File: cppcache/integration-test/fw_dunit.cpp [hunk 3]
```diff
@@ -753,13 +756,22 @@ class TestSlave {
           DUNIT->setSlaveState(m_sId, SLAVE_STATE_TASK_ACTIVE);
           try {
             task->doTask();
+            if (task->m_isHeapAllocated) {
```
Comment (pivotal-jbarrett) [inline cppcache/integration-test/fw_dunit.cpp]: Could we just make it a `std::shared_ptr` and let it do its magic?

File: cppcache/integration-test/testThinClientPartitionResolver.cpp [hunk 4]
```diff
@@ -105,8 +105,8 @@ DUNIT_TASK_DEFINITION(CLIENT1, CreatePoolAndRegions)
     initClient(true);
 
     char endpoints[1024] = {0};
-    sprintf(endpoints, "%s,%s,%s", endpointNames.at(0), endpointNames.at(1),
-            endpointNames.at(2));
+    sprintf(endpoints, "%s,%s,%s", endpointNames.at(0).c_str(),
```
Comment (pivotal-jbarrett) [inline cppcache/integration-test/testThinClientPartitionResolver.cpp]: My not just use `std::string` concatenation?

File: cppcache/include/geode/DataOutput.hpp (hunk 1 above)
Comment (albertogpz) [reply cppcache/include/geode/DataOutput.hpp]: Not sure if this is possible. When the DataOutput constructor is invoked the m_bytes variable is assigned a static buffer which seems to be reused later when other DataOutput instances are created.

File: cppcache/integration-test/fw_dunit.cpp (hunk 3 above)
Comment (albertogpz) [reply cppcache/integration-test/fw_dunit.cpp]: Not really. The framework uses sometimes creates instances in the heap and sometimes in the stack.

File: cppcache/integration-test/testThinClientPartitionResolver.cpp (hunk 4 above)
Comment (albertogpz) [reply cppcache/integration-test/testThinClientPartitionResolver.cpp]: I finally removed this code. It was useless.

File: cppcache/integration-test/ThinClientFailover3.hpp (hunk 2 above)
Comment (albertogpz) [reply cppcache/integration-test/ThinClientFailover3.hpp]: Sure

**Pull request merged**